tonic/functional/event_downsampling.py

```diff
@@ -0,0 +1,271 @@
+    
+    # New event times and indices
+    spike_timestep_pos, index_pos = np.where(np.abs(buffer_layer[first_instance[0]+1:,:,0]) >= buffer_layer_parameters["theta"])
+    spike_timestep_neg, index_neg = np.where(np.abs(buffer_layer[first_instance[1]+1:,:,1]) >= buffer_layer_parameters["theta"])
+    
+    # Restructuring from numpy array to structured array
+    y_pos, x_pos = np.unravel_index(index_pos, np.flip(target_size))
+    y_neg, x_neg = np.unravel_index(index_neg, np.flip(target_size))
+    
+    spike_timestep_pos = (spike_timestep_pos + first_instance[0] + 1) * dt
+    spike_timestep_neg = (spike_timestep_neg + first_instance[1] + 1) * dt
+    
+    events_pos = np.column_stack((x_pos, y_pos, np.ones((len(spike_timestep_pos),1)), spike_timestep_pos.T))
+    events_neg = np.column_stack((x_neg, y_neg, np.zeros((len(spike_timestep_neg),1)), spike_timestep_neg.T))
+    
+    events_new = np.row_stack((events_pos, events_neg))
+    events_new = events_new[events_new[:,-1].argsort()]
+    
+    return unstructured_to_structured(events_new.copy(), dtype=events.dtype)
+
+def differentiator_downsample(events: np.ndarray, sensor_size: tuple, target_size: tuple, dt: float, 
+                              differentiator_time_bins: int = 2, noise_threshold: int = 0):
+    """Downsample using an integrate-and-fire (I-F) neuron model with an additional differentiator 
+    with a noise threshold similar to the membrane potential threshold in the I-F model. Multiply 
+    x/y values by a spatial_factor obtained by dividing sensor size by the target size.
+    Parameters:
+        events (ndarray): ndarray of shape [num_events, num_event_channels].
+        sensor_size (tuple): a 3-tuple of x,y,p for sensor_size.
+        target_size (tuple): a 2-tuple of x,y denoting new down-sampled size for events to be
+                             re-scaled to (new_width, new_height).
+        dt (float): step size for simulation, in ms.
+        differentiator_time_bins (int): number of equally spaced time bins with respect to the dt 
+                                        to be used for the differentiator.
+        noise_threshold (int): number of events before a spike representing a new event is emitted.
+    Returns:
+        the downsampled input events using the differentiator method.
+    """
+        
+    assert "x" and "y" and "t" in events.dtype.names
+    
+    # Create time bins according to differentiator
+    events = time_bin_numpy(events, dt / differentiator_time_bins)
+    
+    # Call integrator method
+    events_integrated = integrator_downsample(events=events, sensor_size=sensor_size, target_size=target_size, 
+                                              noise_threshold=noise_threshold)
+    
+    # All event times
+    events_adjusted = time_bin_numpy(events_integrated, dt)
+    
+    event_times = np.unique(events_adjusted["t"])
+    
+    # Separate by polarity
+    events_positive = events_adjusted[events_adjusted["p"] == 1]
```

Comment:
there is some duplicated code here and in feedback_inhibition, would it be possible to merge that?